### PR TITLE
[Dwarf] Support `__ptrauth` qualifier in metadata nodes

### PR DESCRIPTION
**Pull request by kovdan01**
Emit `__ptrauth`-qualified types as `DIDerivedType` metadata nodes in IR with tag `DW_TAG_LLVM_ptrauth_type`, baseType referring to the type which has the qualifier applied, and the following parameters representing the signing schema:

- `ptrAuthKey` (integer)
- `ptrAuthIsAddressDiscriminated` (boolean)
- `ptrAuthExtraDiscriminator` (integer)
- `ptrAuthIsaPointer` (boolean)
- `ptrAuthAuthenticatesNullValues` (boolean)

### CODE REVIEW

**Comment by kovdan01** [comment]
This applies the following commit https://github.com/ahmedbougacha/llvm-project/commit/857fcbda9529be8d32b357dcf3ac0d2a90319365 with a couple of fixes:
- fix of incorrect usage of `std::optional<bool>` used to store the address discrimination flag https://github.com/access-softek/llvm-project/commit/a34910a6e58f4c441b1ef314898308720542ce6c
- fix of `SubclassData32` usage: it's used for signing schema, but since https://github.com/llvm/llvm-project/commit/102f7fce8d8251655be5bf6955af33e55bea9c9e it's also used for alignment

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 1]
```diff
@@ -1048,6 +1096,48 @@ class DIDerivedType : public DIType {
...
+               ? std::optional<PtrAuthData>(PtrAuthData(SubclassData32))
+               : std::nullopt;
+  }
+
+  /// \returns The PointerAuth key.
+  std::optional<unsigned> getPtrAuthKey() const {
+    if (auto PtrAuthData = getPtrAuthData())
+      return (unsigned)PtrAuthData->Payload.Data.Key;
+    else
+      return std::nullopt;
```
Comment (JDevlieghere) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:1110]: No else after return? Same below.

**Review by JDevlieghere — COMMENTED** [review]
I'm pretty excited about this patch as this will unblock upstreaming (and testing) some of the `ptrauth` debug info attributes.

File: llvm/include/llvm/IR/DebugInfoMetadata.h (hunk 1 above)
Comment (kovdan01) [reply llvm/include/llvm/IR/DebugInfoMetadata.h]: Fixed in 57e86f68e3a83165efa521138b1646ad6a463317, thanks

**Comment by kovdan01** [comment]
@JDevlieghere A kind reminder regarding the PR

**Review by JDevlieghere — APPROVED** [review]
LGTM!

**Pull request merged**